File: OCExample/OCExample/LTPageView/Controller/LTCustomTitleItemView.m
//
//  LTCustomTitleItemView.m
//  OCExample
//
//  Created by gaoliutong on 2022/3/11.
//  Copyright © 2022 LT. All rights reserved.
//

#import "LTCustomTitleItemView.h"
#import "LTScrollView-Swift.h"

/**
 实现步骤：
 1. 指定itemViewClass:
 LTPageView *pageView = [[LTPageView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, H) currentViewController:self viewControllers:viewControllers titles:titles layout:layout itemViewClass:[LTCustomTitleItemView class]];
 
 2. 指定的itemViewClass需要遵守LTPageTitleItemType协议
 
 3. 实现指定的协议方法即可
 */

@interface LTCustomTitleItemView()
@property(strong, nonatomic) UIImageView *backgroundImageView;
@property(strong, nonatomic) UIView *badgeView;
@end

@implementation LTCustomTitleItemView
@synthesize glt_index = _glt_index;

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self insertSubview:self.backgroundImageView atIndex:0];
        [self addSubview:self.badgeView];
    }
    return self;
}

- (void)glt_setSelected:(BOOL)isSelected {
    
    /// 修改背景颜色使用示例
    if (_glt_index % 2 == 0) {
        self.backgroundColor = isSelected ? [UIColor orangeColor] : [UIColor blueColor];
    }else {
        self.backgroundColor = isSelected ? [UIColor greenColor] : [UIColor yellowColor];
    }
    
    /// 设置背景图使用示例
    _backgroundImageView.hidden = _glt_index != 0;
    
    ///选中1的时候隐藏使用示例
    _badgeView.hidden = _glt_index != 1 || isSelected;
    
    /// 设置富文本使用示例
    if (_glt_index == 2) {
        NSMutableAttributedString *attr = [[NSMutableAttributedString alloc] initWithString:self.titleLabel.text];
        [attr setAttributes:@{NSFontAttributeName:[UIFont systemFontOfSize:18], NSForegroundColorAttributeName:[UIColor redColor]} range:NSMakeRange(0, 1)];
        [attr setAttributes:@{NSFontAttributeName:[UIFont systemFontOfSize:10], NSForegroundColorAttributeName:[UIColor blueColor]} range:NSMakeRange(1, 1)];
        self.titleLabel.attributedText = attr;
    }
    
    ///设置标题偏移使用示例
    if (_glt_index == 3) {
        self.titleEdgeInsets = UIEdgeInsetsMake(0, 0, -10, 0);
    }
}

- (void)glt_layoutSubviews {
    self.backgroundImageView.frame = self.bounds;
    self.badgeView.frame = CGRectMake(self.frame.size.width - 15, 5, 10, 10);
}

- (UIImageView *)backgroundImageView {
    if (!_backgroundImageView) {
        _backgroundImageView = [UIImageView new];
        _backgroundImageView.image = [UIImage imageNamed:@"test"];
    }
    return _backgroundImageView;
}

- (UIView *)badgeView {
    if (!_badgeView) {
        _badgeView = [UIView new];
        _badgeView.backgroundColor = [UIColor redColor];
    }
    return _badgeView;
}

- (void)dealloc {
    NSLog(@"LTCustomTitleItemView - dealloc");
}

@end

File: Example/LTScrollView/Demo/LTPageView/Controller/LTCustomTitleItemView.swift
//
//  LTCustomTitleItemView.swift
//  LTScrollView_Example
//
//  Created by gaoliutong on 2022/3/13.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import UIKit

class LTCustomTitleItemView: UIButton, LTPageTitleItemType {
    
    var glt_index: Int = 0
    
    private lazy var backgroundImageView: UIImageView = {
        let imageView = UIImageView()
        imageView.image = UIImage(named: "test")
        imageView.contentMode = .scaleAspectFill
        imageView.layer.masksToBounds = true
        return imageView
    }()
    
    private lazy var badgeView: UIView = {
        let badgeView = UIView()
        badgeView.backgroundColor = .red
        return badgeView
    }()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        insertSubview(backgroundImageView, at: 0)
        addSubview(badgeView)
    }
    
    ///MARK: 协议方法
    func glt_layoutSubviews() {
        backgroundImageView.frame = bounds
        addCornerLayer()
    }
    
    ///MARK: 协议方法
    func glt_setSelected(_ isSelected: Bool) {
        
        print("index = \(glt_index)")
        
        /// 修改背景颜色使用示例
        if glt_index % 2 == 0 {
            backgroundColor = isSelected ? .orange : .blue
        }else {
            backgroundColor = isSelected ? .green : .yellow
        }
        
        /// 设置背景图使用示例
        backgroundImageView.isHidden = glt_index != 0
        
        ///选中1的时候隐藏使用示例
        badgeView.isHidden = glt_index != 1 || isSelected;
        
        /// 设置富文本使用示例
        testFullText()
        
        ///设置标题偏移使用示例
        if glt_index == 3 {
            titleEdgeInsets = UIEdgeInsets(top: 0, left: 0, bottom: -10, right: 0)
        }
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

extension LTCustomTitleItemView {
    
    private func testFullText() {
        guard glt_index == 2 else { return }
        let itemTitle = titleLabel?.text ?? ""
        let attr = NSMutableAttributedString(string: itemTitle)
        attr.addAttributes([NSAttributedString.Key.font : UIFont.systemFont(ofSize: 10)], range: (itemTitle as NSString).range(of: "我是"))
        attr.addAttributes([NSAttributedString.Key.font : UIFont.systemFont(ofSize: 14), NSAttributedString.Key.foregroundColor : UIColor.randomColor], range: (itemTitle as NSString).range(of: "富文本"))
        titleLabel?.attributedText = attr
    }
    
    private func addCornerLayer() {
        guard glt_index == 1 else { return }
        badgeView.frame = CGRect(x: glt_width - 15, y: 5, width: 10, height: 10)
        let maskPath = UIBezierPath(roundedRect: badgeView.bounds, byRoundingCorners: .allCorners, cornerRadii: CGSize(width: 5, height: 5))
        let maskLayer = CAShapeLayer()
        maskLayer.frame = badgeView.bounds
        maskLayer.path = maskPath.cgPath
        badgeView.layer.mask = maskLayer
    }
}
